### PR TITLE
feature: Add Oauth2 Password grant type

### PR DESCRIPTION
**Pull request by Cobalt0s**
# Scripts

## Token

Generating tokens for SugarCRM using Oauth2 password grant type:
image

## Proxy

SugarCRM uses `Access Token` that we acquired using previous script and successfuly proxies to emails resource:
image

### CODE REVIEW

File: providers/utils.go [hunk 1]
```diff
@@ -239,6 +244,9 @@ type NewClientParams struct {
 	// If the provider uses auth code, this field must be set.
 	OAuth2AuthCodeCreds *OAuth2AuthCodeParams
 
+	// TODO comment
```
Comment (laurenzlong) [inline providers/utils.go:247]: I think you forgot to update the comment

File: providers/sugarcrm.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package providers
```
Comment (laurenzlong) [inline providers/sugarcrm.go:1]: can you put this file in a different PR? so have one PR for adding password grant and one PR for adding sugarCRM?

File: providers/sugarcrm.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+const SugarCRM Provider = "sugarCRM"
+
+func init() {
+	// 2-legged auth
+	SetInfo(SugarCRM, ProviderInfo{
+		AuthType: Oauth2,
+		BaseURL:  "{{.workspace}}",
+		Oauth2Opts: &Oauth2Opts{
+			GrantType:                 Password,
+			TokenURL:                  "{{.workspace}}/rest/{{.restVersion}}/oauth2/token",
```
Comment (laurenzlong) [inline providers/sugarcrm.go:12]: let's hard code in a restVersion right now, we don't support a way to get the restVersion from the builder currently, so this won't work with our existing system.

File: providers/utils.go [hunk 4]
```diff
@@ -387,6 +397,31 @@ func createOAuth2ClientCredentialsHTTPClient( //nolint:ireturn
...
 
+func createOAuth2PasswordHTTPClient(ctx context.Context, client *http.Client,
+	dbg bool, params *Oauth2PasswordParams,
+) (common.AuthenticatedHTTPClient, error) {
+	options := []common.OAuthOption{
+		common.WithOAuthClient(getClient(client)),
+		common.WithTokenSourceProvider(func() (*oauth2.Token, error) {
+			// we could use refresh token, but it works a bit differently when refresh token was
+			// acquired via oauth2 password grant type.
+			// TODO this is called every time a request is made. Store AT/RT
```
Comment (laurenzlong) [inline providers/utils.go:408]: what's AT/RT? Can you update the comment to make it more clear?

File: providers/types.yaml [hunk 5]
```diff
@@ -86,7 +86,7 @@ components:
       properties:
         grantType:
           type: string
-          enum: [authorizationCode, clientCredentials, PKCE]
+          enum: [authorizationCode, clientCredentials, password, PKCE]
```
Comment (laurenzlong) [inline providers/types.yaml:89]: we're starting to migrate all of our openapi definition to its own repo, can you make the same change in this file also? https://github.com/amp-labs/openapi/blob/main/catalog/catalog.yaml

File: providers/utils.go (hunk 4 above)
Comment (Cobalt0s) [reply providers/utils.go]: sorry I use AT/RT a short version of access token and refresh token

File: providers/sugarcrm.go (hunk 2 above)
Comment (Cobalt0s) [reply providers/sugarcrm.go:1]: Dedicated PR: https://github.com/amp-labs/connectors/pull/602

File: providers/sugarcrm.go (hunk 3 above)
Comment (Cobalt0s) [reply providers/sugarcrm.go:12]: I updated the base URL to match the version and restrict to rest usage. https://github.com/amp-labs/connectors/pull/602/commits/c8a11f40ac59c98aa88fcf2442aabf895e27a1e4

File: providers/types.yaml (hunk 5 above)
Comment (Cobalt0s) [reply providers/types.yaml:89]: PR created: https://github.com/amp-labs/openapi/pull/4

**Review by eberle1080 — APPROVED** [review]
Please resolve the issues pointed out by @laurenzlong but otherwise this looks good. Thanks! This is very helpful. We won't support this immediately (in the server and frontend) but we can start planning for it. And it does unlock some additional providers.
